tutorials/python/python-type-driven-development/tutorial.md

```diff
@@ -0,0 +1,192 @@
+---
+title: Type Driven Development in Python
+published: 2019-08-01
+tags: Python, typing, types, tdd
+language: python
+author-name: Carlos Villavicencio
+twitter-profile: po5i
+github-profile: po5i
+description: Learn how to improve the software quality and readability by using type driven development in Python.
+---
+
+In a [previous article][types-vs-tests], we discussed the benefits of working with Type Driven Development and the importance of static type checking in compilation time for strongly typed languages such as Java or Haskell. It doesn’t matter if their paradigm is object-oriented or functional,static types add a layer of quality, allows us to reduce the amount of tests we need to write, and reduces our dependency on tests.
+
+Type Driven Development can increase the performance, accuracy, and readability of your code by taking advantage of the language’s type system. There are many books that you can use to learn how to achieve this, for example, “Type-Driven Development with Idris” ([Manning][tdd-idris]) and extrapolate these practices to other languages.
+
+## What about Python?
+
+Is it possible to use a strong type system in a language like Python? The answer is yes. [PEP-484][pep-484] introduced _Type Annotations to Python_ 3.5 back in 2014. It’s been a while, right? Why this isn’t popular? Because Python doesn’t require the code has types everywhere to run, they’re completely optional and mostly used by editors and IDEs. Another reason is that Python code is well-known to be readable in the least number of possible lines of code, but adding types to your code doesn’t make it more readable? What a paradox!
+
+In this article, we’ll be walking through the Python type system, how to add them to our code and Docstrings and perform static type checking using [mypy][mypy]. To start using mypy, install it using pip globally or in your virtual environment.
```

Comment:
@po5i `perform static type checking` I'm curious about `mypy` is it a static analysis tool, a Python superset, or just a library that provide type-checking at runtime? According to the website:

> Compile-time type checking
> Static typing makes it easier to find bugs with less debugging.

So I think it is similar to TS in that regard? I think it would be interesting to compare it with other similar tools out in the development ecosystem. What do you think?

Reply:
@sestrella so `mypy` falls in the first category, a static analysis library. It's not a superset because the type annotations are provided by Python natively. Neither the type-checking at runtime, the type-checking is being executed independently (that's why it's useful for a CI task for example).

I think it's a good idea to elaborate a bit more about the purpose/description of the `mypy` tool. Will work on it.

Reply:
Tutorial updated. @sestrella Would you mind to check it again when you have some time?